### PR TITLE
End-to-end testing decision

### PR DESCRIPTION
**Pull request by 4ydan**
A quick draft of the current frontend testing strategy/stack.

Let me know what you think.
Open for discussion, alternatives, suggestions etc.

@markus2330 @Bushuo @absurd-turtle @ThorbenStaufer

I also moved some stuff around in doc/tests.
Please tag anyone I forgot to tag from the frontend team.

## Basics

- [ ] I added a line to [/doc/CHANGELOG.md](/doc/CHANGELOG.md)
- [ ] The PR is rebased with current master.
- [ ] Details of what you changed are in commit messages.
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildserver is happy.

## Checklist

- [ ] I have installed and I am using [pre-commit hooks](../doc/contrib/README.md#Hooks)
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added unit tests for my code
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](https://github.com/ElektraInitiative/PermaplanT/blob/master/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
      (exceptions are documented)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Examples are well chosen and understandable

### CODE REVIEW

**Review by Bushuo — COMMENTED** [review]
Very reasonable and well thought out decision :)
Let me know if something changes, I will look over it again.

**Comment by 4ydan** [comment]
Thanks for showing me playwright, without you I wouldn't have considered it.

File: doc/decisions/frontend_end_to_end_testing.md
Comment (markus2330) [inline doc/decisions/frontend_end_to_end_testing.md:4]:
```suggestion

```

File: doc/decisions/frontend_end_to_end_testing.md
Comment (markus2330) [inline doc/decisions/frontend_end_to_end_testing.md:14]: :+1: but can you give more details what is needed that this is possible? Are there maybe react specific quirks?

File: doc/decisions/frontend_end_to_end_testing.md
Comment (markus2330) [inline doc/decisions/frontend_end_to_end_testing.md:20]: I don't think this is given, we should create (short) guidelines, see #626

File: doc/decisions/frontend_end_to_end_testing.md
Comment (markus2330) [inline doc/decisions/frontend_end_to_end_testing.md:86]: Please stick to template (see doc/decisions/EXPLANATIONS.md)

```suggestion
## Solutions
```

File: doc/decisions/frontend_end_to_end_testing.md
Comment (markus2330) [inline doc/decisions/frontend_end_to_end_testing.md:77]: Not really needed, we already have a constraint we use BDD and POM.

```suggestion
```

File: doc/decisions/frontend_end_to_end_testing.md
Comment (markus2330) [inline doc/decisions/frontend_end_to_end_testing.md:154]: First introduce solutions, this text is then decision and rationale.

File: doc/decisions/frontend_end_to_end_testing.md
Comment (markus2330) [inline doc/decisions/frontend_end_to_end_testing.md:160]: These are very poor reasons. Please also add technical comparisons. Probably better if you use itemize.

File: doc/decisions/frontend_end_to_end_testing.md
Comment (markus2330) [inline doc/decisions/frontend_end_to_end_testing.md:15]:
```suggestion
4. Should work with Python due to personal preference.

```

File: doc/decisions/frontend_end_to_end_testing.md
Comment (markus2330) [inline doc/decisions/frontend_end_to_end_testing.md:171]: Wrong Link.

File: doc/tests/README.md
Comment (markus2330) [inline doc/tests/README.md:69]: Please move other changes to another PR(s) to keep this focused on the decision.

File: doc/decisions/frontend_end_to_end_testing.md
Comment (markus2330) [inline doc/decisions/frontend_end_to_end_testing.md:169]:
```suggestion
- Pure Javascript, so it doesn't fulfill our constraint of Python.
- Unfamiliarity within our team (including Daniel)
- Some technical limitations (reported by Daniel)
```

**Review by markus2330 — CHANGES_REQUESTED** [review]
Thanks for starting writing the decision. I am missing a few things we discussed:

- [ ] which BDD language and language tests are written in
- [x] Robot Framework should be mentioned as solution (Disclaimer: I am familiar with it but I am very open to another solutions, too)
- [ ] more technical comparison for each of the solutions, e.g. support for model-based? classes? views? features? headless?

File: doc/decisions/frontend_end_to_end_testing.md
Comment (4ydan) [reply doc/decisions/frontend_end_to_end_testing.md:20]: I don't want to duplicate information that is already given on other websites. Maybe a link each that redirects to a guide would be enough?

File: doc/decisions/frontend_end_to_end_testing.md
Comment (4ydan) [reply doc/decisions/frontend_end_to_end_testing.md:14]: BDD and POM support is just required by the testing tools/libraries. BDD is a more difficult one, because the framework needs to parse the feature files, POM you can achieve anywhere where you have classes and objects.

Comment (markus2330) [reply doc/decisions/frontend_end_to_end_testing.md:14]: My comment actually only refers to "Fast and easy prototyping of the tests". So the "fast" and "easy" somehow should be better defined. What makes a framework "fast" or "easy"?

Comment (markus2330) [reply doc/decisions/frontend_end_to_end_testing.md:14]: Maybe thinking of Opensesame helps: we probably would use a different framework there (more likely Robot Framework). What are the differences in the constraints?

Comment (4ydan) [reply doc/decisions/frontend_end_to_end_testing.md:14]: I understand now, I will explain it a bit better.
I'm currently working on the alternative solution robot-framework. Soon I will push my thoughts.

**Comment by 4ydan** [comment]
I hope I could explain myself a little clearer, sorry for not sticking to the template, I saw some other decision have variations of it, but in general, the flow should make more sense now.

I hope `We want to have .feature files and stick to the Gherkin syntax.` answers implicitely the question about the bdd language.
In the next iteration I could add more `technical` comparisons.

**Comment by markus2330** [comment]
> .feature files

What are .features files? Please provide a link (in the decision).

> We are not looking for anything specific like multi browser, device testing support

Probably we need to discuss this more. Device testing support might be a thing, we definitely want Tablet support (and also have it tested)

> performance testing

Yeah, for this we already decided, at least for the tests we have up to now. Let us see if @badnames will continue using this framework (probably yes). Anyway, it is out of scope for this decision.

> Is it enough to mention it like that for all solutions?

I think we didn't look enough what we actually need. Best is if you explicitly say in the assumptions what you think is not a constraint. If I see some misunderstanding (like multi browser, device testing support maybe is), I'll raise my hands :raised_hands:

File: doc/decisions/frontend_end_to_end_testing.md
Comment (markus2330) [inline doc/decisions/frontend_end_to_end_testing.md:12]: Please add links to each.

File: doc/decisions/frontend_end_to_end_testing.md
Comment (markus2330) [inline doc/decisions/frontend_end_to_end_testing.md:18]: What is simple: Lines of codes? Complexity? Availability of good tutorials?

**Review by markus2330 — CHANGES_REQUESTED** [review]
Great job, looks much better! :heart:

File: doc/decisions/frontend_end_to_end_testing.md
Comment (aidnurs) [inline doc/decisions/frontend_end_to_end_testing.md:39]: Hey guys,

leaving a small comment with my thoughts regarding the language here.

Playwright and Python is a great combination, which I’ve used in most of my previous projects. There is a reason why I chose JS + Playwright for PermaplanT for the scrapers, because of the tech stack. Since most of the team knew JS, I tried to avoid introducing an extra language to the stack. This could take extra time for them to switch between languages. 

However, there is no difference in the functionality that I’ve experienced myself.
The situation and the circumstances may have changed after a couple of months, of course, so the final decision is up to you! Hopefully, this helps you with the choice!

File: doc/decisions/frontend_end_to_end_testing.md
Comment (4ydan) [reply doc/decisions/frontend_end_to_end_testing.md:68]: Then I will remove it from the constraint.

File: doc/decisions/frontend_end_to_end_testing.md
Comment (4ydan) [reply doc/decisions/frontend_end_to_end_testing.md:77]: I mean we basically only use the browser from playwright for scraping. I wouldn't really mention that.

Comment (markus2330) [reply doc/decisions/frontend_end_to_end_testing.md:73]: I read that this is considered as advantage of playwright over selenium. Did you notice any difference? :rocket: 

Daniel now also says he is okay with playwright and Python. So we hopefully can finish this decision soon. As you are developing with the right tool anyway, it shouldn't be a problem. Anyway, the next time please create the decision first. :wink:

Comment (4ydan) [reply doc/decisions/frontend_end_to_end_testing.md:73]: In order to make the decision, I had to also play around a bit with the tools (Selenium, Playwright). I noticed that playwright was way easier to install and run. Especiall the browser driver is well better integrated into python than the selenium one.
(I will add this to the decision : ) )

**Comment by 4ydan** [comment]
Ready to merge if this corresponds to your ideas.

File: doc/decisions/frontend_e2e_testing.md
Comment (markus2330) [inline doc/decisions/frontend_e2e_testing.md:47]:
```suggestion
The Robot Framework, while supporting Python and following BDD principles, has its own syntax and separate resource file management, which contradicts the first constraint of the project.
On top of that it is not really used in the fields of frontend testing, even though it has a separate [web testing library](https://robotframework.org/SeleniumLibrary/), which could cause problems when it comes to finding ressources on the Internet.
```

File: doc/decisions/frontend_e2e_testing.md
Comment (markus2330) [inline doc/decisions/frontend_e2e_testing.md:33]: Can you point me to something about this license change? It looks to me it uses Apache 2.0?

File: doc/decisions/frontend_e2e_testing.md
Comment (markus2330) [inline doc/decisions/frontend_e2e_testing.md:30]: The prefix "Alternative" is not really necessary? (It is used several times.)

```suggestion
### [Java](https://www.java.com/de/) + [Selenium](https://www.selenium.dev)
```

**Review by markus2330 — APPROVED** [review]
Decision by itself is fine, only some minor textual improvements and one question (which won't have impact on the outcome).

**Pull request merged**
